File: Python Snippets/radix_sort.py
def radix_sort(nums):
  highest_number = max(nums)
  highest_position = len(str(highest_number))

  buckets = {"0" : [], "1" : [], "2" : [],
             "3" : [], "4" : [], "5" : [],
             "6" : [], "7" : [], "8" : [],
             "9" : []}

  for position in range(1, highest_position + 1):
    helper = []

    for number in nums:
      number_string = str(number)

      if len(number_string) < position:
        buckets["0"].append(number)
      else:
        buckets[number_string[-position]].append(number)

    for bucket in buckets:
      helper += buckets[bucket] # Fill helper array
      buckets[bucket] = [] # Emptying buckets

    nums = helper

  return nums

sample_array = [randint(0, 100) for i in range(20)]
print(f"Unsorted array: {sample_array}")

sorted_array = radix_sort(sample_array)
print(f"Sorted array: {sorted_array}")

File: R Snippets/radix_sort.r
radix_sort <- function(nums) {
  highest_num <- max(nums)
  highest_pos <- nchar(highest_num)

  buckets <- list()

  # create bucket dict
  for (key in 0:9) {
    buckets[[as.character(key)]] <- vector()
  }

  helper <- nums

  for (position in 1:highest_pos) {
    # fill buckets
    for (number in helper) {
      # get the index starting from the right of the number
      index <- nchar(number) - position + 1
      char <- substr(number, index, index)

      if (index <= 0) {
        buckets[["0"]] <- append(buckets[["0"]], number)
      } else {
        buckets[[char]] <- append(buckets[[char]], number)
      }
    }

    # reset helper vector
    helper <- vector()

    # empty buckets
    for (key in names(buckets)) {
      helper <- append(helper, buckets[[key]])
      buckets[[key]] <- vector()
    }
  }

  return (helper)
}

sample_array <- sample(0:50, 25, replace=TRUE)
cat("Unsorted Array:", sample_array, "\n")

sorted_array <- radix_sort(sample_array)
cat("Sorted Array:  ", sorted_array, "\n")
